### PR TITLE
[#111] User can search services by provider

### PR DESCRIPTION
**Pull request by martaswiatkowska**
User can filtr services per provider by click on provider name link.
Link is available on services list and services details view.

### CODE REVIEW

File: app/models/provider.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class Provider < ApplicationRecord
+  has_many :provided_services,
```
[the remark this answers is not in the document]
Comment (mkasztelnik) [reply app/models/provider.rb]: Than this relation can be defined simply:

```ruby
has_many :services, dependent: :nullify
```

File: app/models/service.rb [hunk 2]
```diff
@@ -13,11 +13,15 @@ class Service < ApplicationRecord
   belongs_to :owner,
              class_name: "User",
              optional: true
+  belongs_to :provider,
+             class_name: "Provider",
```
Comment (mkasztelnik) [inline app/models/service.rb]: This is not needed when from the relation name class can be guessed. So here something like this should be enough:

```ruby
belongs_to :provider, optional: true
```

File: app/models/service.rb [hunk 3]
```diff
@@ -13,11 +13,15 @@ class Service < ApplicationRecord
...
+             optional: true
 
   validates :title, presence: true
   validates :description, presence: true
   validates :tagline, presence: true
   validates :connected_url, presence: true, url: true, if: :open_access?
+  validates :provider_id, presence: true
```
Comment (mkasztelnik) [inline app/models/service.rb]:
```ruby
validates :provider, presence: true
```

File: app/views/backoffice/services/_form.html.haml [hunk 4]
```diff
@@ -6,6 +6,7 @@
   = f.input :tagline
   = f.input :connected_url, label: "Service website"
   = f.input :open_access, as: :boolean
+  = f.input :provider_id, collection: get_providers_list
```
Comment (mkasztelnik) [inline app/views/backoffice/services/_form.html.haml]: Take a look at `f.association` simple form helper.

File: app/views/services/_service.html.haml [hunk 5]
```diff
@@ -6,8 +6,8 @@
       .mb-3= service.tagline
       %p.small.float-left.mr-4
         Provided by
-        %a{ href: "" }
-          SHAPE
+        = link_to get_provider_name(service), provider_path(service.provider_id)
```
Comment (mkasztelnik) [inline app/views/services/_service.html.haml]:
```haml
= link_to get_provider_name(service), service.provider
```

File: app/views/services/show.html.haml [hunk 6]
```diff
@@ -4,6 +4,10 @@
   %p
     %b= @service.tagline
   %p= @service.description
+  .row
+    .col
+      Provided by
+      = link_to get_provider_name(@service), provider_path(@service.provider_id)
```
Comment (mkasztelnik) [inline app/views/services/show.html.haml]:
```haml
= link_to get_provider_name(service), service.provider
```

File: config/breadcrumbs.rb [hunk 7]
```diff
@@ -40,3 +40,8 @@
     parent :services
   end
 end
+
+crumb :provider do |provider|
+  link "Provider ##{provider.id}", provider_path(provider)
```
Comment (mkasztelnik) [inline config/breadcrumbs.rb]: Do we really want to show DB id as provider name?

File: lib/tasks/dev.rake [hunk 8]
```diff
@@ -18,9 +18,11 @@ if Rails.env.development?
                        owner: users.sample,
                        open_access: Faker::Boolean.boolean,
                        rating: Random.rand(5.0),
-                       connected_url: Faker::Internet.url)
+                       connected_url: Faker::Internet.url,
+                       provider_id: Provider.all.sample.id)
```
Comment (mkasztelnik) [inline lib/tasks/dev.rake]:
```ruby
provider: all_providers.sample
```

File: lib/tasks/dev.rake [hunk 9]
```diff
@@ -18,9 +18,11 @@ if Rails.env.development?
                        owner: users.sample,
                        open_access: Faker::Boolean.boolean,
                        rating: Random.rand(5.0),
-                       connected_url: Faker::Internet.url)
+                       connected_url: Faker::Internet.url,
```
Comment (mkasztelnik) [inline lib/tasks/dev.rake]: How this is connected with this PR?

File: spec/features/backoffice/services_spec.rb [hunk 10]
```diff
@@ -40,14 +40,19 @@
   end
 
   scenario "I can create new service" do
+    create(:provider)
     visit backoffice_services_path
     click_on "Create new service"
 
+    expect(Provider.all.count).to eq(1)
```
Comment (mkasztelnik) [inline spec/features/backoffice/services_spec.rb]: How is this relevant to this test?

File: spec/helpers/service_helper_spec.rb [hunk 11]
```diff
@@ -14,4 +14,15 @@
   it "converts from decimal 0.0 value to html" do
     expect(print_rating_stars(0.0)).to match(/<i class="far fa-star"><\/i><i class="far fa-star"><\/i><i class="far fa-star"><\/i><i class="far fa-star"><\/i><i class="far fa-star"><\/i>/)
   end
+
+  it "return provider name" do
+    provider = create(:provider)
+    service = create(:service, provider_id: provider.id)
```
Comment (mkasztelnik) [inline spec/helpers/service_helper_spec.rb]: I propose to be explicit when possible:
```ruby
provider = create(:provider, name: "my provider")
service = create(service, provider: provider)
expect(get_provider_name(service)).to eq("my provider")
```

File: lib/tasks/dev.rake (hunk 9 above)
Comment (martaswiatkowska) [reply lib/tasks/dev.rake]: Comma is needed :)

File: lib/tasks/dev.rake (hunk 8 above)
Comment (martaswiatkowska) [reply lib/tasks/dev.rake]: Are you sure? all_providers are defined in seeds.rb

Comment (mkasztelnik) [reply lib/tasks/dev.rake]: You are right the goal here is not to query db every time you are creating service. You can fetch providers list before `times` block.
